File: src/main/java/aug/week1/DetectCapital.java
package aug.week1;

/**
 * Created by Gurupad Mamadapur on 02/08/20
 */
public class DetectCapital {

    public static boolean solve(String word) {
        final int length = word.length();
        if (length == 0) {
            return true;
        }

        char firstChar = word.charAt(0);

        if (isCapital(firstChar)) {
            if (!isCapital(word.charAt(length - 1))) {
                return !anyCapital(word.substring(1));
            } else {
                return allCapital(word);
            }
        } else {
            // make sure the next ones are not
            for (int i = 1; i < length; i++) {
                if (isCapital(word.charAt(i))) {
                    return false;
                }
            }
            return true;
        }
    }

    public static Boolean anyCapital(String word) {
        for (int i = 0; i < word.length(); i++) {
            if (isCapital(word.charAt(i))) {
                return true;
            }
        }
        return false;
    }

    public static Boolean allCapital(String word) {
        for (int i = 0; i < word.length(); i++) {
            if (!isCapital(word.charAt(i))) {
                return false;
            }
        }
        return true;
    }

    public static Boolean isCapital(char s) {
        return s >= 'A' && s <= 'Z';
    }
}

File: src/main/kotlin/aug/week1/DetectCapital.kt
package aug.week1

/**
 * Created by Gurupad Mamadapur on 02/08/20
 */
object DetectCapital {
    fun solve(word: String): Boolean {
        val length = word.length
        if (length == 0) {
            return true
        }
        val firstChar = word[0]
        return if (isCapital(firstChar)) {
            if (!isCapital(word[length - 1])) {
                !word.substring(1).any { isCapital(it) }
            } else {
                word.substring(1).all { isCapital(it) }
            }
        } else {
            // make sure the next ones are not
            word.substring(1).all { !isCapital(it) }
        }
    }

    private fun isCapital(s: Char): Boolean {
        return s in 'A'..'Z'
    }
}